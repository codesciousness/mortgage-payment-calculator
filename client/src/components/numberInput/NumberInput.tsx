import React from 'react';
import TextField from '@mui/material/TextField';
import InputAdornment from '@mui/material/InputAdornment';
import Label from '../label/Label';
import { styles } from '../../styles';

type NumberInputProps = {
    name: string;
    value: number;
    sign: 'dollar' | 'percent';
    width?: number | string;
    dual?: boolean;
    error?: string;
    onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
};

const NumberInput = ({ name, value, sign, width, dual, error, onChange }: NumberInputProps): JSX.Element => {
    const id = name.replaceAll(' ', '');

    const dollarSign = {
        startAdornment: <InputAdornment position="start"><b>$</b></InputAdornment>
    };

    const percentSign = {
        endAdornment: <InputAdornment position="end"><b>%</b></InputAdornment>
    };

    const dollarBorders = {
        borderRadius: '4px 0 0 4px',
        borderRightWidth: 'thin'
    };

    const percentBorders = {
        borderRadius: '0 4px 4px 0',
        borderLeftWidth: 'thin'
    };

    const changeBorders = () => {
        if (dual) {
            if (sign === 'dollar') {
                return dollarBorders;
            }
            else {
                return percentBorders;
            }
        }
        return;
    };

    return (
        <div style={!dual ? styles.input : undefined}>
            {!dual && <Label name={name}/>}
            <TextField
                id={id}
                name={name}
                value={value}
                onChange={onChange}
                helperText={error}
                size='small'
                InputProps={sign === 'dollar' ? dollarSign : percentSign}
                sx={{
                    width,
                    '& .MuiOutlinedInput-root': {
                        '&.Mui-focused fieldset': {
                            borderColor: 'DeepSkyBlue',
                        }
                    },
                    '& fieldset': changeBorders()
                }}
            >
                {value}
            </TextField>
        </div>
    );
};

export default NumberInput;