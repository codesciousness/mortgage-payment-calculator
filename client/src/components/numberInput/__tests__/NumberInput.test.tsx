import React from 'react';
import { render, screen } from '@testing-library/react';
import { create } from 'react-test-renderer';
import NumberInput from '../NumberInput';

const name = 'Label Name';
const value = 100;
const dollarSign = 'dollar';
const percentSign = 'percent'
const width = 300;
const labelTrue = true;
const labelFalse = false;
const dualTrue = true;
const error = 'Error!'
const onChange = () => null;

<NumberInput name={name} value={value} sign={dollarSign} 
    width={width} label={labelTrue} dual={dualTrue}
    onChange={onChange}/>

describe('<NumberInput />', () => {
    test('NumberInput component is defined', () => {
        expect(<NumberInput name={name} value={value} sign={dollarSign} onChange={onChange}/>).toBeDefined();
    });
    test('component matches the snapshot', () => {
        const component = create(<NumberInput name={name} value={value} 
            sign={dollarSign} width={width} label={labelTrue} dual={dualTrue}
            onChange={onChange}/>);
        expect(component.toJSON()).toMatchSnapshot();
    });
    test('renders correctly when sign prop is dollar', () => {
        render(<NumberInput name={name} value={value} sign={dollarSign} onChange={onChange}/>);
        expect(screen.queryByText('Label Name')).not.toBeInTheDocument();
        expect(screen.getByDisplayValue('100')).toBeInTheDocument();
        expect(screen.getByText('$')).toBeInTheDocument();
    });
    test('renders correctly when sign prop is percent', () => {
        render(<NumberInput name={name} value={value} sign={percentSign} onChange={onChange}/>);
        expect(screen.queryByText('Label Name')).not.toBeInTheDocument();
        expect(screen.getByDisplayValue('100')).toBeInTheDocument();
        expect(screen.getByText('%')).toBeInTheDocument();
    });
    test('renders correctly when label prop is true', () => {
        render(<NumberInput name={name} value={value} sign={dollarSign} label={labelTrue} onChange={onChange}/>);
        expect(screen.getByText('Label Name')).toBeInTheDocument();
        expect(screen.getByDisplayValue('100')).toBeInTheDocument();
    });
    test('renders correctly when label prop is false', () => {
        render(<NumberInput name={name} value={value} sign={dollarSign} label={labelFalse} onChange={onChange}/>);
        expect(screen.queryByText('Label Name')).not.toBeInTheDocument();
        expect(screen.getByDisplayValue('100')).toBeInTheDocument();
    });
    test('renders correctly when error prop is provided', () => {
        render(<NumberInput name={name} value={value} sign={dollarSign} error={error} onChange={onChange}/>);
        expect(screen.getByDisplayValue('100')).toBeInTheDocument();
        expect(screen.getByText('Error!')).toBeInTheDocument();
    });
});