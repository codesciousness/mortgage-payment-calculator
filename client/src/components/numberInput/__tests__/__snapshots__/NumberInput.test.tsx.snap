// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NumberInput /> component matches the snapshot 1`] = `
<div
  className="NumberInput"
>
  <label
    className="MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-root MuiFormLabel-colorPrimary css-22ecsy-MuiFormLabel-root-MuiInputLabel-root"
    id="LabelName"
  >
    Label Name
  </label>
  <div
    className="MuiFormControl-root MuiTextField-root css-171j53h-MuiFormControl-root-MuiTextField-root"
  >
    <div
      className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedStart css-1q6at85-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <div
        className="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeSmall css-ittuaa-MuiInputAdornment-root"
      >
        <span
          className="notranslate"
        >
          ​
        </span>
        <b>
          $
        </b>
      </div>
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedStart css-1ua80n0-MuiInputBase-input-MuiOutlinedInput-input"
        disabled={false}
        id="LabelName"
        name="Label Name"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="text"
        value={100}
      />
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-hdw1oc"
        >
          <span
            className="notranslate"
          >
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
